tasks/Task.cpp

```diff
@@ -2,48 +2,422 @@
+                mState.peer_connection.state = Disconnected;
+            case rtc::PeerConnection::State::Closed:
+                mState.peer_connection.state = Closed;
+            case rtc::PeerConnection::State::Connected:
+                mState.peer_connection.state = Connected;
+            case rtc::PeerConnection::State::Connecting:
+                mState.peer_connection.state = Connecting;
+            case rtc::PeerConnection::State::Failed:
+                mState.peer_connection.state = Failed;
+            case rtc::PeerConnection::State::New:
+                mState.peer_connection.state = NewConnection;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onGatheringStateChange(
+        [&](rtc::PeerConnection::GatheringState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::GatheringState::Complete:
+                mState.peer_connection.gathering_state = Complete;
+            case rtc::PeerConnection::GatheringState::InProgress:
+                mState.peer_connection.gathering_state = InProgress;
+            case rtc::PeerConnection::GatheringState::New:
+                mState.peer_connection.gathering_state = NewGathering;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onSignalingStateChange(
+        [&](rtc::PeerConnection::SignalingState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::SignalingState::HaveLocalOffer:
+                mState.peer_connection.signaling_state = HaveLocalOffer;
+            case rtc::PeerConnection::SignalingState::HaveLocalPranswer:
+                mState.peer_connection.signaling_state = HaveLocalPranswer;
+            case rtc::PeerConnection::SignalingState::HaveRemoteOffer:
+                mState.peer_connection.signaling_state = HaveRemoteOffer;
+            case rtc::PeerConnection::SignalingState::HaveRemotePranswer:
+                mState.peer_connection.signaling_state = HaveRemotePranswer;
+            case rtc::PeerConnection::SignalingState::Stable:
+                mState.peer_connection.signaling_state = Stable;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onLocalDescription(
+        [&, remote_peer_id](rtc::Description description)
+        {
+            Json::Value message;
+            message["protocol"] = "one-to-one";
+            message["to"] = remote_peer_id;
+            message["action"] = description.typeString();
+            message["data"]["from"] = _local_peer_id.get();
+            message["data"]["description"] = string(description);
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+            {
+                Json::FastWriter fast;
+                ws->send(fast.write(message));
+                mState.peer_connection.local_description = NewDescription;
```

Comment:
This handling of mState's local_description is misleading. The false branch is not that there was no local description, it is that you did not send it.

Tbh, I think it's irrelevant here. If the websocket is being destroyed, it means we're shutting down the component and it's fine to not send anything.

(This applies to the other flags dependent on write/no-write)